src/persist-txn/src/operator.rs

```diff
@@ -228,191 +296,105 @@ where
+                // output to `logical_upper`. This also indicates that we need
+                // to wait for the next DataRemapEntry. It can either (A) have
+                // the same physical upper or (B) have a larger physical upper.
+                //
+                // - If (A), then we would again satisfy this `physical_upper`
+                //   check, again advance the logical upper again, ...
+                // - If (B), then we'd fall down to the code below, which copies
+                //   the passthrough data until the frontier passes
+                //   `physical_upper`, then loops back up here.
+                if r.physical_upper.less_equal(cap.time()) {
+                    if cap.time() < &r.logical_upper {
+                        cap.downgrade(&r.logical_upper);
                     }
+                    remap = txns_progress_frontiers_read_remap_input(
+                        &name,
+                        &mut remap_input,
+                        r.clone(),
+                    )
+                    .await;
+                    continue;
                 }
             }
 
-            // Any time we hit this point, we've emitted everything known to be
-            // physically written to the data shard. Query the txns shard to
-            // find out what to do next given our current progress.
-            loop {
-                txns_cache.update_ge(&output_progress_exclusive).await;
-                txns_cache.compact_to(&output_progress_exclusive);
-                let data_listen_next = txns_cache
-                    .state
-                    .data_listen_next(&data_id, output_progress_exclusive.clone());
-                debug!(
-                    "{} {:.9} data_listen_next at {:?}({:?}): {:?}",
-                    name,
-                    data_id.to_string(),
-                    read_data_to.elements(),
-                    output_progress_exclusive,
-                    data_listen_next,
-                );
-                match data_listen_next {
-                    // We've caught up to the txns upper and we have to wait for
-                    // it to advance before asking again.
+            // This only returns None when there are no more data left. Turn it
+            // into an empty frontier progress so we can re-use the shutdown
+            // code below.
+            let event = passthrough_input
+                .next()
+                .await
+                .unwrap_or_else(|| Event::Progress(Antichain::new()));
+            match event {
+                // NB: Ignore the data_cap because this input is disconnected.
+                Event::Data(_data_cap, mut data) => {
+                    // NB: Nothing to do here for `until` because the both
+                    // `shard_source` (before this operator) and
+                    // `mfp_and_decode` (after this operator) do the necessary
+                    // filtering.
```

Comment:
```suggestion
                    // NB: Nothing to do here for `until` because both the
                    // `shard_source` (before this operator) and
                    // `mfp_and_decode` (after this operator) do the necessary
                    // filtering.
```

Reply:
Done!